File: python/lis_dp_n2.py
import sys
from typing import Optional

from utils import load_sequence_from_json


def lis_dp(sequence: list[int]) -> list[int]:
    """
    O(n^2) DP algorithm that returns one Longest Increasing Subsequence.
    Time: O(n^2), Space: O(n)
    """
    n: int = len(sequence)
    if n == 0:
        return []

    # dp[i] = length of LIS ending at i
    dp: list[int] = [1] * n
    prev: list[Optional[int]] = [None] * n

    for i in range(n):
        for j in range(i):
            if sequence[j] < sequence[i] and dp[j] + 1 > dp[i]:
                dp[i] = dp[j] + 1
                prev[i] = j

    # find index of maximum dp
    max_len: int = max(dp)
    max_idx: int = dp.index(max_len)

    # reconstruct LIS
    lis: list[int] = []
    idx: Optional[int] = max_idx
    while idx is not None:
        lis.append(sequence[idx])
        idx = prev[idx]
    lis.reverse()
    return lis


def main() -> None:
    if len(sys.argv) < 2:
        print("Usage: uv run python/lis_naive.py <N>")
        sys.exit(1)

    n = int(sys.argv[1])
    file_path = f"data/sequence_{n}.json"

    try:
        sequence = load_sequence_from_json(file_path)
    except FileNotFoundError:
        print(f"Error: File '{file_path}' not found.")
        print("Run 'uv run python/generator.py' to generate the data.")
        sys.exit(1)

    result = lis_dp(sequence)

    print(f"LIS length = {len(result)}")
    # print(f"LIS = {result}")


if __name__ == "__main__":
    main()

File: rust/lis_dp_n2.rs
//! ```cargo
//! [dependencies]
//! serde = { version = "1.0", features = ["derive"] }
//! serde_json = "1.0"
//! ```

use std::error::Error;
use std::fs;

/// Load a sequence from JSON (same format as Python).
fn load_sequence(path: &str) -> Result<Vec<i32>, Box<dyn Error>> {
    let content = fs::read_to_string(path)?;
    let sequence: Vec<i32> = serde_json::from_str(&content)?;
    Ok(sequence)
}

/// Compute the Longest Increasing Subsequence using dynamic programming (O(n^2)).
fn lis_dp(sequence: &[i32]) -> Vec<i32> {
    let n = sequence.len();
    if n == 0 {
        return vec![];
    }

    let mut dp: Vec<usize> = vec![1; n];
    let mut prev: Vec<Option<usize>> = vec![None; n];

    for i in 0..n {
        for j in 0..i {
            if sequence[j] < sequence[i] && dp[j] + 1 > dp[i] {
                dp[i] = dp[j] + 1;
                prev[i] = Some(j);
            }
        }
    }

    // find index of maximum dp
    let mut max_idx = 0usize;
    for i in 1..n {
        if dp[i] > dp[max_idx] {
            max_idx = i;
        }
    }

    // reconstruct LIS
    let mut lis_rev: Vec<i32> = Vec::new();
    let mut cur: Option<usize> = Some(max_idx);
    while let Some(idx) = cur {
        lis_rev.push(sequence[idx]);
        cur = prev[idx];
    }
    lis_rev.reverse();
    lis_rev
}

fn main() -> Result<(), Box<dyn Error>> {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: rust-script rust/lis_dp_n2.rs <N>");
        std::process::exit(1);
    }

    let n_arg = &args[1];
    let n: u32 = n_arg.replace('_', "").parse().unwrap();
    let file_path = format!("data/sequence_{}.json", n);

    let sequence = match load_sequence(&file_path) {
        Ok(seq) => seq,
        Err(_) => {
            eprintln!("Error: file '{}' does not exist.", file_path);
            eprintln!("Run 'uv run python/generator.py' to generate the data.");
            std::process::exit(1);
        }
    };

    let lis = lis_dp(&sequence);
    println!("LIS length = {}", lis.len());
    // println!("LIS = {:?}", lis);

    Ok(())
}
